File: arkts_static_test/03_Types/ets_cases/3.12_Type_null/cross_lang_verify/TYP_03_12_Type_null.java
// 001: Null literal - Java allows null only for reference types
class TYP_03_12_001_NullLiteral {
    public static void main(String[] args) {
        String x = null;
        Object y = null;
        System.out.println("TYP_03_12_001 verified: x=" + x + ", y=" + y);
    }
}

// 002: Nullable int - Java uses Integer wrapper
class TYP_03_12_002_NullableInt {
    public static void main(String[] args) {
        Integer x = null;
        x = 42;
        x = null;
        Integer y = 100;
        y = null;
        System.out.println("TYP_03_12_002 verified: x=" + x + ", y=" + y);
    }
}

// 003: Nullable String
class TYP_03_12_003_NullableString {
    public static void main(String[] args) {
        String s = null;
        s = "hello";
        s = null;
        System.out.println("TYP_03_12_003 verified: s=" + s);
    }
}

// 004: Nullable Object
class TYP_03_12_004_NullableObject {
    public static void main(String[] args) {
        Object o = null;
        o = new Object();
        o = null;
        System.out.println("TYP_03_12_004 verified: o=" + o);
    }
}

// 005: Triple union - Java has no T|null|undefined
class TYP_03_12_005_TripleUnion {
    public static void main(String[] args) {
        Object x = null;
        x = 1;
        x = null;
        System.out.println("TYP_03_12_005 verified: x=" + x);
    }
}

// 006: Function returning null
class TYP_03_12_006_FunctionReturnNull {
    static String maybeNull(int v) {
        if (v > 0) return "positive";
        return null;
    }
    public static void main(String[] args) {
        String r1 = maybeNull(1);
        String r2 = maybeNull(-1);
        System.out.println("TYP_03_12_006 verified: r1=" + r1 + ", r2=" + r2);
    }
}

// 007: Function parameter with null
class TYP_03_12_007_FunctionParamNull {
    static int acceptNullable(String p) {
        if (p == null) return -1;
        return p.length();
    }
    public static void main(String[] args) {
        acceptNullable(null);
        acceptNullable("hello");
        System.out.println("TYP_03_12_007 verified");
    }
}

// 008: Null in switch - Java cannot switch on null, must use if-else
class TYP_03_12_008_NullSwitch {
    static String classifyValue(String v) {
        if (v == null) return "is_null";
        switch (v) {
            case "": return "is_empty";
            default: return "is_string";
        }
    }
    public static void main(String[] args) {
        String r1 = classifyValue(null);
        String r2 = classifyValue("");
        String r3 = classifyValue("hello");
        System.out.println("TYP_03_12_008 verified: " + r1 + ", " + r2 + ", " + r3);
    }
}

// 009: Null assigned to Object - KEY DIFF: Java allows, ArkTS forbids for nullish
class TYP_03_12_009_NullAny {
    public static void main(String[] args) {
        Object a = null;
        a = null;
        a = 42;
        System.out.println("TYP_03_12_009 verified: a=" + a);
    }
}

// 013 FAIL equivalent: Java allows null for reference types (KEY DIFFERENCE from ArkTS)
class TYP_03_12_013_NullAssignNonNullish {
    public static void main(String[] args) {
        String s = null;  // Java: OK, but ArkTS: compile-fail
        Object o = null; // Java: OK, but ArkTS: compile-fail
        // int n = null;  // Java: compile error - int is primitive
        System.out.println("TYP_03_12_013 DIFFERENCE: Java allows null for reference types");
    }
}

// 016 FAIL equivalent: Java allows nullish -> Object (KEY DIFFERENCE)
class TYP_03_12_016_NullishToObject {
    public static void main(String[] args) {
        Object nullable = null;
        Object o = nullable; // Java: OK, but ArkTS: compile-fail
        System.out.println("TYP_03_12_016 DIFFERENCE: Java allows null to Object");
    }
}

// 022 RUNTIME: Null type narrowing
class TYP_03_12_022_RuntimeNarrowing {
    static int safeStringLength(String s) {
        if (s == null) return -1;
        return s.length();
    }
    public static void main(String[] args) {
        int r1 = safeStringLength(null);
        if (r1 != -1) throw new RuntimeException("null should return -1, got " + r1);
        int r2 = safeStringLength("hello");
        if (r2 != 5) throw new RuntimeException("hello length should be 5, got " + r2);
        int r3 = safeStringLength("");
        if (r3 != 0) throw new RuntimeException("empty length should be 0, got " + r3);
        System.out.println("TYP_03_12_022 verified");
    }
}

// 025 RUNTIME: Null in conditional
class TYP_03_12_025_RuntimeSwitch {
    static String classifyValue(String v) {
        if (v == null) return "is_null";
        switch (v) {
            case "": return "is_empty";
            default: return "is_string";
        }
    }
    public static void main(String[] args) {
        String r1 = classifyValue(null);
        if (!r1.equals("is_null")) throw new RuntimeException("expected is_null");
        String r2 = classifyValue("");
        if (!r2.equals("is_empty")) throw new RuntimeException("expected is_empty");
        String r3 = classifyValue("hello");
        if (!r3.equals("is_string")) throw new RuntimeException("expected is_string");
        System.out.println("TYP_03_12_025 verified");
    }
}

File: arkts_static_test/03_Types/ets_cases/3.12_Type_null/cross_lang_verify/TYP_03_12_Type_null.swift
// 001: Null literal - Swift uses nil for Optional types
let x001: String? = nil
let y001: String? = nil
print("TYP_03_12_001 verified: x=\(x001 ?? "nil"), y=\(y001 ?? "nil")")

// 002: Nullable int - Swift uses Int? (Optional)
var x002: Int? = nil
x002 = 42
x002 = nil
var y002: Int? = 100
y002 = nil
print("TYP_03_12_002 verified: x=\(x002 ?? -1), y=\(y002 ?? -1)")

// 003: Nullable String
var s003: String? = nil
s003 = "hello"
s003 = nil
print("TYP_03_12_003 verified: s=\(s003 ?? "nil")")

// 004: Nullable Object - Swift uses AnyObject?
class MyClass004 {}
var o004: AnyObject? = nil
o004 = MyClass004()
o004 = nil
print("TYP_03_12_004 verified: o=\(o004 == nil ? "nil" : "object")")

// 005: Triple union - Swift has no T|null|undefined; uses Any?
var x005: Any? = nil
x005 = 1
x005 = nil
print("TYP_03_12_005 verified: x=\(x005 ?? "nil" as Any)")

// 006: Function returning nil
func maybeNull(v: Int) -> String? {
    if v > 0 {
        return "positive"
    }
    return nil
}
let r006a: String? = maybeNull(v: 1)
let r006b: String? = maybeNull(v: -1)
print("TYP_03_12_006 verified: r1=\(r006a ?? "nil"), r2=\(r006b ?? "nil")")

// 007: Function parameter with nil
func acceptNullable(p: String?) -> Int {
    if p == nil { return -1 }
    return p!.count
}
let _007a = acceptNullable(p: nil)
let _007b = acceptNullable(p: "hello")
print("TYP_03_12_007 verified: \(_007a), \(_007b)")

// 008: Nil in switch - Swift uses case nil for Optional
func classifyValue(_ v: String?) -> String {
    switch v {
    case nil: return "is_nil"
    case "": return "is_empty"
    default: return "is_string"
    }
}
let r008a = classifyValue(nil)
let r008b = classifyValue("")
let r008c = classifyValue("hello")
print("TYP_03_12_008 verified: \(r008a), \(r008b), \(r008c)")

// 009: Nil assigned to Any - Swift: Any? can hold nil
var a009: Any? = nil
a009 = nil
a009 = 42
print("TYP_03_12_009 verified: a=\(a009 ?? "nil" as Any)")

// 013 DIFFERENCE: Swift does NOT allow nil for non-Optional types
// let n013: Int = nil      // compile error
// let s013: String = nil   // compile error
// Swift matches ArkTS behavior here (compile error for nil on non-optional)
// But Java allows nil for reference types (key difference)
print("TYP_03_12_013 DIFFERENCE: Swift does not allow nil for non-Optional types (matches ArkTS)")

// 016 DIFFERENCE: Swift requires unwrapping Optional before assigning to non-Optional
// let o016: AnyObject = Optional<AnyObject>.none  // compile error
// Swift matches ArkTS: cannot assign nil/Optional to non-Optional
print("TYP_03_12_016 DIFFERENCE: Swift requires unwrapping (matches ArkTS nullish incompatibility)")

// 022 RUNTIME: Nil type narrowing
func safeLength(_ s: String?) -> Int {
    if s == nil { return -1 }
    return s!.count
}
let r022a = safeLength(nil)
if r022a != -1 { fatalError("null input should return -1, got \(r022a)") }
let r022b = safeLength("hello")
if r022b != 5 { fatalError("'hello' length should be 5, got \(r022b)") }
let r022c = safeLength("")
if r022c != 0 { fatalError("'' length should be 0, got \(r022c)") }
print("TYP_03_12_022 verified")

// 025 RUNTIME: Nil in conditional (Swift uses if-let / switch nil)
func classify025(_ v: String?) -> String {
    if v == nil { return "is_nil" }
    switch v! {
    case "": return "is_empty"
    default: return "is_string"
    }
}
let r025a = classify025(nil)
if r025a != "is_nil" { fatalError("expected is_nil, got \(r025a)") }
let r025b = classify025("")
if r025b != "is_empty" { fatalError("expected is_empty, got \(r025b)") }
let r025c = classify025("hello")
if r025c != "is_string" { fatalError("expected is_string, got \(r025c)") }
print("TYP_03_12_025 verified")